### PR TITLE
(ISA-275) Optionally remember saved state

### PR DESCRIPTION
**Pull request by JoshGx2000**
[ISA-275](https://jira.its.utas.edu.au/browse/ISA-275)

So I'm not certain on my implementation, which is why I'm looking to you for a second opinion.
The way I've done this is I've registered a new event named `:put-hash-if-autosave`, and everything that previously used the `:put-hash` effect now dispatches that event instead.

I was considering registering a new effect for this, but decided against it due to clunkiness (since it would have to check if saves were enabled, encode the state, and put-hash in a single effect – [day8 discourage complex effects like that](https://github.com/day8/re-frame/blob/master/docs/Effects.md#:~:text=A%20word%20of%20advice%20-%20make%20them%20as%20simple%20as%20possible%2C%20and%20then%20simplify%20them%20further.%20You%20don%27t%20want%20them%20containing%20any%20fancy%20logic.)).

### CODE REVIEW

**Review by markhepburn — CHANGES_REQUESTED** [review]
This is more about a discussion than anything else: if I'm reading this correctly, the unspoken implication is that you'll no longer be able to paste links?  You'd have to use the share-button?
(that makes me think about other side-effects, like what happens when you load a saved-state link, and does that load a url-hash which then doesn't change?)

Might be worth drafting an email for this one to discuss implications.

File: database/Procedures/UpdateBathymetry.sql [hunk 1]
```diff
@@ -3,95 +3,108 @@
...
 
+-- The @batch optional parameter dictates where in the procedure we should
+-- begin/continue (for the event in which the procedure fails partway through
+-- execution). A @batch value of 0 indicates we start at the beginning, values 1-5
+-- are for the different stages of geometry aggregation, a value of 6 is for
+-- updating the BathymetryGeoms tables, and a value of 7 updates the hardcoded
+-- bathymetry boundaries tables. Smaller @batch values execute larger @batch value
+-- operations automatically (i.e. starting with @batch of 3 will execute 3 as well
+-- as all the @batch=4 stuff, @batch=4 executes 4 and @batch=5 stuff, and so on).
+
```
[the remark this answers is not in the document]
Comment (JoshGx2000) [reply database/Procedures/UpdateBathymetry.sql:14]: Ah yep, whoops

File: frontend/src/cljs/imas_seamap/events.cljs [hunk 2]
```diff
@@ -395,13 +395,11 @@
...
-     (when (seq habitat-layers) ; only display transect plot if there's an active habitat layer
-       {:dispatch-n [[:transect.plot/show]
-                     [:transect.query/habitat    query-id linestring]
-                     [:transect.query/bathymetry query-id linestring]]}))))
-
+     {:db         db
+      :dispatch-n [[:put-hash-if-autosave]
+                   (when (seq habitat-layers) [:transect.plot/show])                                 ; only display transect plot if there's an active habitat layer
+                   (when (seq habitat-layers) [:transect.query/habitat query-id linestring])         ; only query transect habitat
+                   (when (seq habitat-layers) [:transect.query/bathymetry query-id linestring])]}))) ; only query transect bathymetry
```
[the remark this answers is not in the document]
Comment (JoshGx2000) [reply frontend/src/cljs/imas_seamap/events.cljs:402]: Yep, that seems much better

**Comment by JoshGx2000** [comment]
> This is more about a discussion than anything else: if I'm reading this correctly, the unspoken implication is that you'll no longer be able to paste links? You'd have to use the share-button? (that makes me think about other side-effects, like what happens when you load a saved-state link, and does that load a url-hash which then doesn't change?)
> 
> Might be worth drafting an email for this one to discuss implications.

My understanding of what was said is that we only want to update the url-hash if autosave is enabled. I'm getting this from the ticket where it's said that a page refresh would return the map to the state it was in when the page was loaded.

I don't see where you're getting the no longer being able to paste hash-state links thing from. Can you explain?

File: frontend/src/cljs/imas_seamap/core.cljs [hunk 3]
```diff
@@ -101,6 +102,8 @@
     :create-save-state                    [events/create-save-state]
     :create-save-state-success            [events/create-save-state-success]
     :create-save-state-failure            [events/create-save-state-failure]
+    :toggle-autosave-application-state    [events/toggle-autosave-application-state]
+    :put-hash-if-autosave                 [events/put-hash-if-autosave]
```
[the remark this answers is not in the document]
Comment (JoshGx2000) [reply frontend/src/cljs/imas_seamap/core.cljs]: Renamed!

File: frontend/src/cljs/imas_seamap/events.cljs (hunk 2 above)
[the remark this answers is not in the document]
Comment (JoshGx2000) [reply frontend/src/cljs/imas_seamap/events.cljs]: Fixed!

**Comment by markhepburn** [comment]
> My understanding of what was said is that we only want to update the url-hash if autosave is enabled. I'm getting this from the ticket where it's said that a page refresh would return the map to the state it was in when the page was loaded.
> 
> I don't see where you're getting the no longer being able to paste hash-state links thing from. Can you explain?

Their usual workflow is "create map, select url bar, copy link".  If for eg, they open a link they've been sent (which would now have to be created from the share button, which is fine if that's consistent), moved the map around but don't have autosave enabled, then copied the url bar again... they'd be sending the same link.

Basically I'm trying to move away from "literally what was in the ticket" (with a smidgeon of assumptions about implementation details), into what they might have been intending, and why they want to disable autosave.

**Comment by JoshGx2000** [comment]
Ah, okay. I'll bring it up with Emma then and see what she says (or if she can ask Peter Walsh who originally requested the feature).

**Comment by JoshGx2000** [comment]
Updated autosave now that uses cookies

**Review by markhepburn — APPROVED** [review]

**Comment by markhepburn** [comment]
Nice job -- merge conflicts needing fixing :+1:

**Pull request merged**